gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -138,12 +134,12 @@ public void shutdown() {
 
   @Override
   public boolean isShutdown() {
-    return executor.isShutdown();
+    return future.isCancelled();
   }
 
   @Override
   public boolean isTerminated() {
-    return executor.isTerminated();
+    return future.isDone();
```

Comment:
Probably equivalent to `future.isCancelled()`.

Reply:
I think isDone and isCancelled are equivalent. And both are slightly wrong for isTerminated. I think the strict definition of isTerminated is: not only has the task been unscheduled, but also that the last invocation stopped running.

Practically I dont think it matters too much. So if you want to take a shortcut and just use isDone, but document it. 

Alternatively use a CountdownLatch where each run() is wrapped in an increment & decrement operation on the latch. awaitTermination can then block on the latch

Reply:
From the Javadoc for isDone:
> Returns true if this task completed. Completion may be due to normal termination, an exception, or cancellation -- in all of these cases, this method will return true.

It sounds like the last invocation would have stopped running. What makes you think otherwise?

Reply:
I think the cancellation of the future is only cancelling the scheduled tasks, if the `run()` method is already running and takes long time to complete, then `isDone()` could return true while the `run()` is still running. I just reproduced this in my local as well. 
That being said, not sure if it matters that much in practice like Igor said, I don't know how long it takes for a typical `run()` to complete. Using a CountdownLatch is not enough in this case since it does not support increment, only decrement. So we may need something more if we want to fix it perfectly.